internal/session.go

```diff
@@ -0,0 +1,421 @@
+// on the normal taskList (as user specified in ActivityOptions) and may be picked up by another worker since
+// it's not in a session.
+//
+// This API will return an error if the session in the context has already completed, or the resource release
+// process has failed. No error will returned if user tries to complete a session that has already failed and
+// the API call won't do anything.
+func CompleteSession(ctx Context) error {
+	sessionInfo := getSessionInfo(ctx)
+	if sessionInfo == nil || sessionInfo.sessionState == sessionStateClosed {
+		return errNoOpenSession
+	}
+
+	if sessionInfo.sessionState == sessionStateFailed {
+		return nil
+	}
+	retryPolicy := &RetryPolicy{
+		InitialInterval:    time.Second,
+		BackoffCoefficient: 1.1,
+		MaximumInterval:    time.Second * 10,
+		MaximumAttempts:    5,
+	}
+	ao := ActivityOptions{
+		ScheduleToStartTimeout: time.Second * 10,
+		StartToCloseTimeout:    time.Second * 10,
+		RetryPolicy:            retryPolicy,
+	}
+	completionCtx := WithActivityOptions(ctx, ao)
+	// the tasklist will be overrided to use the one stored in sessionInfo
+	err := ExecuteActivity(completionCtx, sessionCompletionActivityName, sessionInfo.SessionID).Get(ctx, nil)
+	if err == nil {
+		sessionInfo.sessionState = sessionStateClosed
+	} else {
+		sessionInfo.sessionState = sessionStateFailed
+	}
+	return err
+}
+
+// GetSessionInfo returns the sessionInfo stored in the context. If there are multiple sessions in the context,
+// (for example, the same context is used to create, complete, create another session. Then user found that the
+// session has failed, and created a new one on it), the most recent sessionInfo will be returned.
+//
+// This API will return an error if there's no sessionInfo in the context.
+func GetSessionInfo(ctx Context) (SessionInfo, error) {
+	info := getSessionInfo(ctx)
+	if info == nil {
+		return SessionInfo{}, errNoSessionInfo
+	}
+	return *info, nil
+}
+
+func getSessionInfo(ctx Context) *SessionInfo {
+	info := ctx.Value(sessionInfoContextKey)
+	if info == nil {
+		return nil
+	}
+	return info.(*SessionInfo)
+}
+
+func setSessionInfo(ctx Context, sessionInfo *SessionInfo) Context {
+	return WithValue(ctx, sessionInfoContextKey, sessionInfo)
+}
+
+func createSession(ctx Context, creationTasklist string, retryable bool) (Context, error) {
+	if prevSessionInfo := getSessionInfo(ctx); prevSessionInfo != nil && prevSessionInfo.sessionState == sessionStateOpen {
+		return nil, errFoundExistingOpenSession
+	}
+	sessionID, err := generateSessionID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tasklistChan := GetSignalChannel(ctx, sessionID) // use sessionID as channel name
+	// Retry is only needed when creating new session and the error returned is NewCustomError(errTooManySessionsMsg)
+	retryPolicy := &RetryPolicy{
+		InitialInterval:    time.Second,
+		BackoffCoefficient: 1.1,
+		MaximumInterval:    time.Second * 10,
+		MaximumAttempts:    10,
```

Comment:
Make retryExpiration configurable. Do not specify maxAttempts.

Reply:
Done. Use creationTimeout as expirationInterval.